8250989: Consolidate buffer allocation code for CDS static/dynamic dumping

iklam opened the pull request:
Before this PR, when a static CDS archive is dumped, we would unconditionally allocate a 4GB ReservedSpace and use that as the buffer for writing the archive content. We usually don't need such a big ReservedSpace. this also unnecessarily complicates Metaspace::global_initialize().

After this PR, both we first load all the classes and then allocate a buffer with an appropriate size.

I also simplified the pointer relocation code that prepares the archive to be mapped at the "requested location" (usually 0x800000000), and improved the comments.

We used to have lots of special cases for adjusting pointers during dynamic dump. All of those are removed.

Reviewers, please start with
- Header comments in archiveBuilder.hpp that describe the overall process of archive dumping
- `MetaspaceShared::initialize_for_static_dump()`
- `ArchiveBuilder::reserve_buffer()`
- Comments around `RelocateBufferToRequested` in archiveBuilder.cpp

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8250989](https://bugs.openjdk.java.net/browse/JDK-8250989): Consolidate buffer allocation code for CDS static/dynamic dumping


### Reviewers
 * [Calvin Cheung](https://openjdk.java.net/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to eac503514cab096ddd4475b642ba39509171373c
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to eac503514cab096ddd4475b642ba39509171373c


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2296/head:pull/2296`
`$ git checkout pull/2296`

iklam commented:
/label remove hotspot
/label add hotspot-runtime

calvinccheung commented on src/hotspot/share/memory/archiveBuilder.cpp:
```diff
@@ -613,9 +712,124 @@ void ArchiveBuilder::make_klasses_shareable() {
+  return requested_p - _requested_static_archive_bottom;
+}
+
+uintx ArchiveBuilder::any_to_offset(address p) const {
+  if (is_in_mapped_static_archive(p)) {
+    return p - _mapped_static_archive_bottom;
+  }
+  return buffer_to_offset(p);
+}
```
Is it possible to add an `assert(DynamicDumpSharedSpaces)` after the `if` statement at line 728?

iklam replied:
Added per suggestion.

calvinccheung commented on test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/DynamicArchiveRelocationTest.java:
```diff
@@ -97,11 +93,7 @@ static void doTest() throws Exception {
-            out.shouldContain("ArchiveRelocationMode == 1: always allocate class space at an alternative address");
-            out.shouldContain("Relocating archive from");
-        }
+        OutputAnalyzer out = TestCommon.dumpBaseArchive(baseArchiveName, unlockArg, logArg);
```
Should the check for `out.shouldContain("Relocating archive from")` be kept?
Otherwise, no need to save the `OutputAnalyzer` returned from `TestCommon.dumpBaseArchive`.

iklam replied:
I added the `out.shouldContain("Relocating archive from")` back (so that the changes are the same as in ArchiveRelocationTest.java)

calvinccheung commented on src/hotspot/share/classfile/symbolTable.cpp:
```diff
@@ -609,15 +607,6 @@ void SymbolTable::write_to_archive(GrowableArray<Symbol*>* symbols) {
-    // Verify the written shared table is correct -- at this point,
-    // vmSymbols has already been relocated to point to the archived
-    // version of the Symbols.
-    Symbol* sym = vmSymbols::java_lang_Object();
-    const char* name = (const char*)sym->bytes();
-    int len = sym->utf8_length();
-    unsigned int hash = hash_symbol(name, len, _alt_hash);
-    assert(sym == _shared_table.lookup(name, hash, len), "sanity");
```
Maybe keeping this within a DEBUG_ONLY block if it's still applicable?

iklam replied:
This check cannot be done anymore -- `_shared_table` is initialized with the "requested" addresses but `sym`is still at "buffer" address, so the check will fail (I forgot whether the lookup would crash or the assert will fail).

calvinccheung reviewed (commented):
The dynamic dumping code is cleaner after this change. Looks good.
I've a few minor comments.

tstuefe commented on src/hotspot/share/memory/metaspace.cpp:
```diff
@@ -672,20 +672,23 @@ void Metaspace::global_initialize() {
+
   // If UseCompressedClassPointers=1, we have two cases:
-  // a) if CDS is active (either dump time or runtime), it will create the ccs
+  // a) if CDS archive has been mapped, it will create the ccs
```
This comment is confusing. So, at this point the ccs archive has already be mapped? Where?

iklam replied:
I've changed it to:

```
  // a) if CDS archive(s) have been mapped, we would have already reserved the
  //    ccs (immediately above the mapped archives) and initialized
  //    the CompressedKlassPointers encoding.
```

tstuefe replied:
No, the confusing part is the order. The old comment used future tense, this uses past tense. So to the reader it looks like mapping may already have happened. But at this point in the code the archive has not been mapped yet, no?
Proposal:
```
  // If UseCompressedClassPointers=1, we have two cases:
  // a) if CDS is active (runtime, Xshare=on), it will create the class space
  //    for us, initialize it and set up CompressedKlassPointers encoding.
  //    Class space will be reserved above the mapped archives.
  // b) if CDS either deactivated (Xshare=off) or a static dump is to be done (Xshare:dump),
  //    we will create the class space on our own. It will be placed above the java heap,
  //    since we assume it has been placed in low
  //    address regions. We may rethink this (see JDK-8244943). Failing that,
  //    it will be placed anywhere.
```

iklam replied:
This looks good. I added it to the code.

tstuefe commented on src/hotspot/share/memory/metaspaceShared.cpp:
```diff
@@ -201,149 +201,29 @@ static char* compute_shared_base(size_t cds_total) {
-    //  must be allocated near the cds such as that the compressed Klass pointer
-    //  encoding can be used to en/decode pointers from both cds and ccs. Since
-    //  Metaspace cannot do this (it knows nothing about cds), we do it for
-    //  Metaspace here and pass it the space to use for ccs.
```
So this whole comment is wrong now and obsolete?

iklam replied:
Yes, most of the work done under the old function `MetaspaceShared::initialize_dumptime_shared_and_meta_spaces()` are now unnecessary. Specifically, it doesn't initialize the "shared space" or "meta space" anymore. That's why I renamed it to `MetaspaceShared::initialize_for_static_dump()`. 

Now it only does the following:
- reset `SharedBaseAddress` if too high
- reserve `_symbol_rs`

===============
BTW, I have added comments at the top of archiveBuilder.hpp to give an overview of how the memory is reserved and relocated during the dump process.

tstuefe replied:
Thanks for the clarification.

tstuefe reviewed (commented):
Hi Ioi,

Just some initial remarks. Had a cursory glance at this patch, but this is a very large change :) seems to be a good cleanup though.

I am trying to understand how allocation works now. Just to make sure I get the gist right, where before we had three cases (with UseCompressedClassPoitners):
1) DumpSharedSpaces = cds+ccs in a 4G range, realized by just reserving 4g
2) UseSharedSpaces = cds+ccs in a 4G range, realized with our complicated mapping dance
3) no sharing, where class space is allocated like in old times

now you eliminated (1)? So the restriction that ccs has to be within narrow klass ptr range of cds does not apply anymore at dump time? And we can map ccs wherever?

You know, at some point a flowchart would be helpful.

tstuefe reviewed (commented):
Hi Ioi,

not a full review, but the metaspace related changes to reservations look fine. See minor nits below.

This is unrelated to your patch: in compute_shared_base, you set SharedBaseAddress in case of an error. I think it should either be just a calculator function with no side effects - just returning the corrected value - or always set SharedBaseAddress and _requested_base_address always, returning void. That would be clearer.

One question, SharedBaseAddress vs MetaspaceShared::_requested_base_address : what is the difference? When do you use one, when the other?

Thanks, Thomas

coleenp commented on src/hotspot/share/memory/archiveBuilder.hpp:
```diff
@@ -161,7 +205,7 @@ class ArchiveBuilder : public StackObj {
 
   // For global access.
-  static ArchiveBuilder* _singleton;
+  static ArchiveBuilder* _current;
```
This is kind of a strange concept since ArchiveBuilder is a StackObj.  Maybe it should just be static if only one thread uses it at a time?

coleenp replied:
And can a different thread call current() ? yikes.

iklam replied:
ArchiveBuilder has temporarily allocated objects such as `ArchiveBuilder::_ptrmap`. These are easily freed when the ArchiveBuilder falls out of scope. For example, in `DynamicArchive::dump()`.

`current()` should be callable only inside the VM thread, where the dumping operations take place. I'll add an assert for that.

coleenp reviewed (approved):
Looks good. Some questions though.

iklam commented on src/hotspot/share/memory/metaspace.cpp:
```diff
@@ -672,22 +672,23 @@ void Metaspace::global_initialize() {
-    MetaspaceShared::initialize_dumptime_shared_and_meta_spaces();
-  } else if (UseSharedSpaces) {
-    // If any of the archived space fails to map, UseSharedSpaces
-    // is reset to false.
```
OK. I added the comments back, but put it below the call to`MetaspaceShared::initialize_runtime_shared_and_meta_spaces();`

iklam commented on src/hotspot/share/oops/oop.cpp:
```diff
@@ -143,6 +143,14 @@ bool oopDesc::has_klass_gap() {
 }
 
+#if INCLUDE_CDS_JAVA_HEAP
+void oopDesc::set_narrow_klass(narrowKlass nk) {
```
I am not sure about that. Friend classes can access all private methods.

iklam commented:
> Hi Ioi,
> 
> not a full review, but the metaspace related changes to reservations look fine. See minor nits below.
> 
> This is unrelated to your patch: in compute_shared_base, you set SharedBaseAddress in case of an error. I think it should either be just a calculator function with no side effects - just returning the corrected value - or always set SharedBaseAddress and _requested_base_address always, returning void. That would be clearer.
>

I changed `compute_shared_base()` to have no side effect.

> One question, SharedBaseAddress vs MetaspaceShared::_requested_base_address : what is the difference? When do you use one, when the other?
> 

I added comments in metaspaceShared.hpp:

```
  // This is the base address as specified by -XX:SharedBaseAddress during -Xshare:dump.
  // Both the base/top archives are written using this as their base address.
  //
  // During static dump: _requested_base_address == SharedBaseAddress.
  //
  // During dynamic dump: _requested_base_address is not always the same as SharedBaseAddress:
  // - SharedBaseAddress is used for *reading the base archive*. I.e., CompactHashtable uses
  //   it to convery offsets to pointers to Symbols in the base archive.
  //   The base archive may be mapped to an OS-selected address due to ASLR. E.g.,
  //   you may have SharedBaseAddress == 0x00ff123400000000.
  // - _requested_base_address is used for *writing the output archive*. It's usually
  //   0x800000000 (unless it was set by -XX:SharedBaseAddress during -Xshare:dump).
  static char* requested_base_address() {
    return _requested_base_address;
  }
```

Thanks!
Ioi

tstuefe commented:
> > Hi Ioi,
> > not a full review, but the metaspace related changes to reservations look fine. See minor nits below.
> > This is unrelated to your patch: in compute_shared_base, you set SharedBaseAddress in case of an error. I think it should either be just a calculator function with no side effects - just returning the corrected value - or always set SharedBaseAddress and _requested_base_address always, returning void. That would be clearer.
> 
> I changed `compute_shared_base()` to have no side effect.
> 
> > One question, SharedBaseAddress vs MetaspaceShared::_requested_base_address : what is the difference? When do you use one, when the other?
> 
> I added comments in metaspaceShared.hpp:
> 
> ```
>   // This is the base address as specified by -XX:SharedBaseAddress during -Xshare:dump.
>   // Both the base/top archives are written using this as their base address.
>   //
>   // During static dump: _requested_base_address == SharedBaseAddress.
>   //
>   // During dynamic dump: _requested_base_address is not always the same as SharedBaseAddress:
>   // - SharedBaseAddress is used for *reading the base archive*. I.e., CompactHashtable uses
>   //   it to convery offsets to pointers to Symbols in the base archive.
>   //   The base archive may be mapped to an OS-selected address due to ASLR. E.g.,
>   //   you may have SharedBaseAddress == 0x00ff123400000000.
>   // - _requested_base_address is used for *writing the output archive*. It's usually
>   //   0x800000000 (unless it was set by -XX:SharedBaseAddress during -Xshare:dump).
>   static char* requested_base_address() {
>     return _requested_base_address;
>   }
> ```
> 
> Thanks!
> Ioi

Much clearer, thanks!

Cheers, Thomas

dcubed-ojdk commented on src/hotspot/share/memory/metaspaceShared.hpp:
```diff
@@ -239,20 +227,27 @@ class MetaspaceShared : AllStatic {
+  //
+  // During dynamic dump: _requested_base_address is not always the same as SharedBaseAddress:
+  // - SharedBaseAddress is used for *reading the base archive*. I.e., CompactHashtable uses
+  //   it to convery offsets to pointers to Symbols in the base archive.
```
Perhaps:
s/convery/convey/

iklam replied:
Thanks for spotting it. It's "convert". I'll fixed it :-)

coleenp commented on src/hotspot/share/memory/archiveBuilder.hpp:
```diff
@@ -348,6 +350,7 @@ class ArchiveBuilder : public StackObj {
   }
 
   static ArchiveBuilder* current() {
+    assert_is_vm_thread();
```
Ok, thank you!

iklam commented:
/integrate
